xskillscore/core/deterministic.py

```diff
@@ -6,14 +6,61 @@
 __all__ = ['pearson_r', 'pearson_r_p_value', 'rmse', 'mse', 'mae']
 
 
-def _preprocess(dim):
+def _preprocess_dims(dim):
+    """Preprocesses dimensions to prep for stacking.
+    
+    Parameters
+    ----------
+    dim : str, list
+        The dimension(s) to apply the function along.
+    """
     if isinstance(dim, str):
         dim = [dim]
     axis = tuple(range(-1, -len(dim) - 1, -1))
     return dim, axis
 
 
-def pearson_r(a, b, dim):
+def _preprocess_weights(a, dim, new_dim, weights):
+    """Preprocesses weights array to prepare for numpy computation.
+    
+    Parameters
+    ----------
+    a : Dataset, DataArray, GroupBy, Variable, numpy/dask arrays or scalars
+        One of the arrays over which the function will be applied.
+    dim : str, list
+        The original dimension(s) to apply the function along.
+    new_dim : str
+        The newly named dimension after running ``_preprocess_dims``
+    weights : Dataset, DataArray, GroupBy, Variable, numpy/dask arrays or scalars
+        Weights to apply to function, matching the dimension size of ``new_dim``.
+    """
+    if weights is None:
+        weights = xr.ones_like(a)  # Evenly weight
```

Comment:
This makes it so users cannot submit an ndarray. I need to find a workaround here. Also should add testing for non-xarray objects.